### PR TITLE
reconciler/workload/resource: remove syncer finalizer when SyncTarget is gone

### LINKED ISSUE #1353: Some workspaces stuck in permanent failure to delete
1. Start kcp
2. Run `go test ./test/e2e/reconciler/scheduling -args --use-default-kcp-server`
3. Wait for test to finish (and pass)
4. Check kcp log

It'll look similar to this:

```
I0624 16:35:35.490027   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root#$#e2e-org-gzzsd
I0624 16:35:35.501072   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-9vcrg
I0624 16:35:35.502764   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-dqz78
I0624 16:35:35.504870   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-h5tf4
I0624 16:35:35.539283   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root#$#e2e-org-gzzsd after 49.271236ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: clusterworkspaces.tenancy.kcp.dev has 3 resource instances; Some content in the workspace has finalizers remaining: tenancy.kcp.dev/workspace-finalizer in 3 resource instances
I0624 16:35:36.414194   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-h5tf4
I0624 16:35:36.414221   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-dqz78
I0624 16:35:36.422587   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/kube-root-ca.crt
I0624 16:35:36.424160   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/kube-root-ca.crt
I0624 16:35:36.428215   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/default-token-phmdn
I0624 16:35:36.428225   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/default-token-jlvrw
I0624 16:35:36.447711   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/default
I0624 16:35:36.448961   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/default
I0624 16:35:36.454205   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-9vcrg
I0624 16:35:36.471272   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/default
I0624 16:35:36.472397   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/syncer-workloadcluster-950756
I0624 16:35:36.474823   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/default
I0624 16:35:36.484208   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root:e2e-org-gzzsd#$#e2e-workspace-h5tf4 after 70.03937ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: configmaps. has 1 resource instances, namespaces. has 1 resource instances, secrets. has 1 resource instances, serviceaccounts. has 1 resource instances; Some content in the workspace has finalizers remaining: workload.kcp.dev/syncer-workloadcluster-950756 in 3 resource instances
I0624 16:35:36.485250   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root:e2e-org-gzzsd#$#e2e-workspace-dqz78 after 71.0512ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: configmaps. has 1 resource instances, namespaces. has 1 resource instances, secrets. has 1 resource instances, serviceaccounts. has 1 resource instances; Some content in the workspace has finalizers remaining: workload.kcp.dev/syncer-workloadcluster-950756 in 3 resource instances
I0624 16:35:36.491635   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/default-token-m9scm
I0624 16:35:36.492438   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/syncer-workloadcluster-950756-token-28c6f
I0624 16:35:36.494191   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/default-token-p6b7k
I0624 16:35:36.494992   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/kubernetes
I0624 16:35:36.499498   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/kube-root-ca.crt
I0624 16:35:36.500990   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/kube-root-ca.crt
I0624 16:35:36.524007   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root:e2e-org-gzzsd#$#e2e-workspace-9vcrg after 69.81841ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: configmaps. has 2 resource instances, namespaces. has 2 resource instances, secrets. has 4 resource instances, serviceaccounts. has 3 resource instances; Some content in the workspace has finalizers remaining: workload.kcp.dev/syncer-workloadcluster-950756 in 9 resource instances
I0624 16:35:38.430783   42042 informer.go:271] Discovering types for logical cluster "root:default"
I0624 16:35:38.435468   42042 informer.go:271] Discovering types for logical cluster "root:e2e-org-gzzsd"
I0624 16:35:38.437979   42042 informer.go:271] Discovering types for logical cluster "root:e2e-org-gzzsd:e2e-workspace-9vcrg"
I0624 16:35:38.440489   42042 informer.go:271] Discovering types for logical cluster "root:e2e-org-gzzsd:e2e-workspace-h5tf4"
I0624 16:35:38.443088   42042 informer.go:271] Discovering types for logical cluster "root:e2e-org-gzzsd:e2e-workspace-dqz78"
I0624 16:35:38.539914   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root#$#e2e-org-gzzsd
I0624 16:35:38.557968   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-9vcrg
I0624 16:35:38.560006   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-dqz78
I0624 16:35:38.561849   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-h5tf4
I0624 16:35:38.593265   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root#$#e2e-org-gzzsd after 53.373963ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: clusterworkspaces.tenancy.kcp.dev has 3 resource instances; Some content in the workspace has finalizers remaining: tenancy.kcp.dev/workspace-finalizer in 3 resource instances
I0624 16:35:39.484557   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-h5tf4
I0624 16:35:39.485517   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-dqz78
I0624 16:35:39.495468   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/kube-root-ca.crt
I0624 16:35:39.502164   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/default
I0624 16:35:39.508882   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/default-token-phmdn
I0624 16:35:39.520490   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/kube-root-ca.crt
I0624 16:35:39.523541   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/default
I0624 16:35:39.524162   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-9vcrg
I0624 16:35:39.529384   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/kube-root-ca.crt
I0624 16:35:39.530140   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/default-token-jlvrw
I0624 16:35:39.531663   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/kube-root-ca.crt
I0624 16:35:39.540281   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/default
I0624 16:35:39.541141   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/syncer-workloadcluster-950756
I0624 16:35:39.543326   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/default
I0624 16:35:39.545565   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root:e2e-org-gzzsd#$#e2e-workspace-h5tf4 after 61.025194ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: configmaps. has 1 resource instances, namespaces. has 1 resource instances, secrets. has 1 resource instances, serviceaccounts. has 1 resource instances; Some content in the workspace has finalizers remaining: workload.kcp.dev/syncer-workloadcluster-950756 in 3 resource instances
I0624 16:35:39.546217   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root:e2e-org-gzzsd#$#e2e-workspace-dqz78 after 60.732022ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: configmaps. has 1 resource instances, namespaces. has 1 resource instances, secrets. has 1 resource instances, serviceaccounts. has 1 resource instances; Some content in the workspace has finalizers remaining: workload.kcp.dev/syncer-workloadcluster-950756 in 3 resource instances
I0624 16:35:39.563825   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/default-token-m9scm
I0624 16:35:39.564647   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/syncer-workloadcluster-950756-token-28c6f
I0624 16:35:39.566273   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/default-token-p6b7k
I0624 16:35:39.567077   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-9vcrg/kcp-system/kubernetes
I0624 16:35:39.590595   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root:e2e-org-gzzsd#$#e2e-workspace-9vcrg after 66.437822ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: configmaps. has 2 resource instances, namespaces. has 2 resource instances, secrets. has 4 resource instances, serviceaccounts. has 3 resource instances; Some content in the workspace has finalizers remaining: workload.kcp.dev/syncer-workloadcluster-950756 in 9 resource instances
I0624 16:35:41.593774   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root#$#e2e-org-gzzsd
I0624 16:35:41.616571   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-9vcrg
I0624 16:35:41.618484   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-dqz78
I0624 16:35:41.620506   42042 store.go:1019] DEBUG: DELETE key func returned: /tenancy.kcp.dev/clusterworkspaces/customresources/root:e2e-org-gzzsd/e2e-workspace-h5tf4
I0624 16:35:41.648043   42042 clusterworkspace_deletion_controller.go:136] Content remaining in workspace root#$#e2e-org-gzzsd after 54.284856ms, waiting 3 seconds to continue: some content remains in the workspace, estimate 5 seconds before it is removed: Some resources are remaining: clusterworkspaces.tenancy.kcp.dev has 3 resource instances; Some content in the workspace has finalizers remaining: tenancy.kcp.dev/workspace-finalizer in 3 resource instances
I0624 16:35:42.546457   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-h5tf4
I0624 16:35:42.546534   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-dqz78
I0624 16:35:42.558007   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/default
I0624 16:35:42.561317   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/kube-root-ca.crt
I0624 16:35:42.565046   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-dqz78/default/default
I0624 16:35:42.590757   42042 clusterworkspace_deletion_controller.go:165] Deleting workspace root:e2e-org-gzzsd#$#e2e-workspace-9vcrg
I0624 16:35:42.594309   42042 store.go:1019] DEBUG: DELETE key func returned: /core/configmaps/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/kube-root-ca.crt
I0624 16:35:42.597257   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/default
I0624 16:35:42.598123   42042 store.go:1019] DEBUG: DELETE key func returned: /core/serviceaccounts/root:e2e-org-gzzsd:e2e-workspace-9vcrg/default/syncer-workloadcluster-950756
I0624 16:35:42.598171   42042 store.go:1019] DEBUG: DELETE key func returned: /core/secrets/root:e2e-org-gzzsd:e2e-workspace-h5tf4/default/default-token-jlvrw
```

The `workload.kcp.dev/syncer-*` finalizers are not being removed. It is currently the responsibility of the syncer to do that. But if the syncer stops working, cluster goes away, etc., we can get stuck in this state.

### PR DESCRIPTION
**Pull request by sttts**
When a SyncTarget is deleted, the syncer loses the ability to remove its finalizer from objects because the VW will not let it see its objects. This PR will make the workload resource controller do this by checking existence of the SyncTargets.

Fixes #1353

### CODE REVIEW

**Comment by ncdc** [comment]
I guess this is related to #1353 that I just filed

File: pkg/indexers/workload.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+)
+
+const (
+	SyncTargetsBySyncTargetKey = "SyncTargetsBySyncTargetKey"
+)
+
+func IndexSyncTargetsBySyncTargetKey(obj interface{}) ([]string, error) {
```
Comment (sttts) [inline pkg/indexers/workload.go:31]: @ncdc is our plan to move all indexers here? I think I would like that.

**Comment by davidfestal** [comment]
FTR this would be done by the Virtual Workspace in the context of the 1st Syncer Virtual Workspace Transformation PR, which manages the deletion process previously managed by the Syncer, and also the cleanup of related annotations / labels.

**Pull request merged**